presto-main/src/main/java/com/facebook/presto/event/QueryProgressMonitor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.event;
+
+import com.facebook.presto.dispatcher.DispatchManager;
+import com.facebook.presto.server.BasicQueryInfo;
+import io.airlift.units.Duration;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.facebook.presto.execution.QueryState.FAILED;
+import static com.facebook.presto.execution.QueryState.FINISHED;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class QueryProgressMonitor
+{
+    private final ScheduledExecutorService queryProgressMonitorExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("query-progress-monitor-executor"));
+    private final QueryMonitor queryMonitor;
+    private final DispatchManager dispatchManager;
+    private final Duration queryProgressPublishInterval;
+
+    @Inject
+    public QueryProgressMonitor(
+            QueryMonitor queryMonitor,
+            DispatchManager dispatchManager,
+            QueryMonitorConfig queryMonitorConfig)
+    {
+        this.queryMonitor = requireNonNull(queryMonitor, "queryMonitor is null");
+        this.dispatchManager = requireNonNull(dispatchManager, "dispatchManager is null");
+        this.queryProgressPublishInterval = requireNonNull(queryMonitorConfig, "queryMonitorConfig is null").getQueryProgressPublishInterval();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        queryProgressMonitorExecutor.scheduleWithFixedDelay(
```

Comment:
For a heavily contended cluster, I think this would be pretty wasteful, because most queries will publish identical information each time even though they're queued and it won't change until they dequeue.  What do you think about making it based off of query state transition?  I.e. it publishes once per query state transition, and when it transitions into EXECUTING then we start a periodic timer?

Reply:
I feel that the decision to process/ignore events for QUEUED queries should be left to the consumer itself. It also depends on the use case for which these events are being utilized. For instance, within Meta itself, these events will be used for two different purposes: 

1) Resource utilization of in-progress queries to perform quota enforcement and throttling. This consumer simply ignore events for QUEUED queries and only logs events for running queries. 

2) Getting the current queuing state of the cluster and power certain alerts/dashboard. This requires getting events for queued queries at periodic interval in order to generate a recent snapshot. Another problem I see is that if a query fails after queueing and before logging query completed event (in case of coordinator crash or cluster restart), then we will not be able to figure out if a query is in QUEUED state because its waiting for resources or whether it died due to to some reason. So, having periodic samples will surface out more accurate snaphot of the cluster state.